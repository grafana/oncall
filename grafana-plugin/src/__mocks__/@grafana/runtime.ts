export const getBackendSrv = () => ({
  get: jest.fn(),
  post: jest.fn(),
});
