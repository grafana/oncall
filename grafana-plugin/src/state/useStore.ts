import React from 'react';

import { MobXProviderContext } from 'mobx-react';

import { RootStore } from './index';

export function useStore(): RootStore {
  const { store } = React.useContext(MobXProviderContext);
  return store;
}
