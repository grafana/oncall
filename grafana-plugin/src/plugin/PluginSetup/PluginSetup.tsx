import React, { FC, PropsWithChildren, useCallback, useEffect } from 'react';

import { PluginPage as RealPluginPage } from '@grafana/runtime'; // Use the one from @grafana, not our wrapped PluginPage
import { Button, HorizontalGroup, LinkButton } from '@grafana/ui';
import { PluginPageFallback } from 'PluginPage';
import { observer } from 'mobx-react';
import { AppRootProps } from 'types';

import logo from 'assets/img/logo.svg';
import { isTopNavbar } from 'plugin/GrafanaPluginRootPage.helpers';
import { useStore } from 'state/useStore';
import { loadJs } from 'utils/loadJs';

export type PluginSetupProps = AppRootProps & {
  InitializedComponent: (props: AppRootProps) => JSX.Element;
};

type PluginSetupWrapperProps = PropsWithChildren<{
  text: string;
}>;

const PluginSetupWrapper: FC<PluginSetupWrapperProps> = ({ text, children }) => {
  const PluginPage = (isTopNavbar() ? RealPluginPage : PluginPageFallback) as React.ComponentType<any>;

  return (
    <PluginPage>
      <div className="spin">
        <img alt="Grafana OnCall Logo" src={logo} />
        <div className="spin-text">{text}</div>
        {children}
      </div>
    </PluginPage>
  );
};

export const PluginSetup: FC<PluginSetupProps> = observer(({ InitializedComponent, ...props }) => {
  const store = useStore();
  const setupPlugin = useCallback(() => store.setupPlugin(props.meta), [props.meta]);

  useEffect(() => {
    (async function () {
      await setupPlugin();
      store.recaptchaSiteKey &&
        loadJs(`https://www.google.com/recaptcha/api.js?render=${store.recaptchaSiteKey}`, store.recaptchaSiteKey);
    })();
  }, [setupPlugin]);

  if (store.initializationError) {
    return (
      <PluginSetupWrapper text={store.initializationError}>
        {!store.currentlyUndergoingMaintenance && (
          <div className="configure-plugin">
            <HorizontalGroup>
              <Button variant="primary" onClick={setupPlugin} size="sm">
                Retry
              </Button>
              <LinkButton href={`/plugins/grafana-oncall-app?page=configuration`} variant="primary" size="sm">
                Configure Plugin
              </LinkButton>
            </HorizontalGroup>
          </div>
        )}
      </PluginSetupWrapper>
    );
  }
  return <InitializedComponent {...props} />;
});
