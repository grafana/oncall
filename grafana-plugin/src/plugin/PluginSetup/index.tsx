import React, { FC, PropsWithChildren, useCallback, useEffect } from 'react';

import { PluginPage as RealPluginPage } from '@grafana/runtime'; // Use the one from @grafana, not our wrapped PluginPage
import { Button, HorizontalGroup, LinkButton } from '@grafana/ui';
import { PluginPageFallback } from 'PluginPage';
import { observer } from 'mobx-react';
import { AppRootProps } from 'types';

import logo from 'img/logo.svg';
import { isTopNavbar } from 'plugin/GrafanaPluginRootPage.helpers';
import { useStore } from 'state/useStore';

export type PluginSetupProps = AppRootProps & {
  InitializedComponent: (props: AppRootProps) => JSX.Element;
};

type PluginSetupWrapperProps = PropsWithChildren<{
  text: string;
}>;

const PluginSetupWrapper: FC<PluginSetupWrapperProps> = ({ text, children }) => {
  const PluginPage = (isTopNavbar() ? RealPluginPage : PluginPageFallback) as React.ComponentType<any>;

  return (
    <PluginPage>
      <div className="spin">
        <img alt="Grafana OnCall Logo" src={logo} />
        <div className="spin-text">{text}</div>
        {children}
      </div>
    </PluginPage>
  );
};

const PluginSetup: FC<PluginSetupProps> = observer(({ InitializedComponent, ...props }) => {
  const store = useStore();
  const setupPlugin = useCallback(() => store.setupPlugin(props.meta), [props.meta]);

  useEffect(() => {
    setupPlugin();
  }, [setupPlugin]);

  if (store.appLoading) {
    return <PluginSetupWrapper text="Initializing plugin..." />;
  }

  if (store.initializationError) {
    return (
      <PluginSetupWrapper text={store.initializationError}>
        <div className="configure-plugin">
          <HorizontalGroup>
            <Button variant="primary" onClick={setupPlugin} size="sm">
              Retry
            </Button>
            <LinkButton href={`/plugins/grafana-oncall-app?page=configuration`} variant="primary" size="sm">
              Configure Plugin
            </LinkButton>
          </HorizontalGroup>
        </div>
      </PluginSetupWrapper>
    );
  }

  return <InitializedComponent {...props} />;
});

export default PluginSetup;
