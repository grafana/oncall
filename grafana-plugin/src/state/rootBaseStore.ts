import { AppPluginMeta } from '@grafana/data';
import { getBackendSrv } from '@grafana/runtime';
import { action, observable } from 'mobx';
import qs from 'query-string';
import { OnCallAppSettings } from 'types';

import { AlertReceiveChannelStore } from 'models/alert_receive_channel/alert_receive_channel';
import { AlertReceiveChannel } from 'models/alert_receive_channel/alert_receive_channel.types';
import { AlertReceiveChannelFiltersStore } from 'models/alert_receive_channel_filters/alert_receive_channel_filters';
import { AlertGroupStore } from 'models/alertgroup/alertgroup';
import { ApiTokenStore } from 'models/api_token/api_token';
import { CloudStore } from 'models/cloud/cloud';
import { EscalationChainStore } from 'models/escalation_chain/escalation_chain';
import { EscalationPolicyStore } from 'models/escalation_policy/escalation_policy';
import { GlobalSettingStore } from 'models/global_setting/global_setting';
import { GrafanaTeamStore } from 'models/grafana_team/grafana_team';
import { HeartbeatStore } from 'models/heartbeat/heartbeat';
import { MaintenanceStore } from 'models/maintenance/maintenance';
import { OrganizationLogStore } from 'models/organization_log/organization_log';
import { OutgoingWebhookStore } from 'models/outgoing_webhook/outgoing_webhook';
import { ResolutionNotesStore } from 'models/resolution_note/resolution_note';
import { ScheduleStore } from 'models/schedule/schedule';
import { SlackStore } from 'models/slack/slack';
import { SlackChannelStore } from 'models/slack_channel/slack_channel';
import { TeamStore } from 'models/team/team';
import { TelegramChannelStore } from 'models/telegram_channel/telegram_channel';
import { UserStore } from 'models/user/user';
import { UserGroupStore } from 'models/user_group/user_group';
import { makeRequest } from 'network';

import { AppFeature } from './features';
import { createGrafanaToken, getPluginSyncStatus, installPlugin, startPluginSync, updateGrafanaToken } from './plugin';
import { UserAction } from './userAction';

// ------ Dashboard ------ //

export class RootBaseStore {
  @observable
  appLoading = true;

  @observable
  backendVersion = '';

  @observable
  backendLicense = '';

  @observable
  pluginIsInitialized = true;

  @observable
  correctProvisioningForInstallation = true;

  @observable
  correctRoleForInstallation = true;

  @observable
  signupAllowedForPlugin = true;

  @observable
  initializationError = '';

  @observable
  retrySync = false;

  @observable
  isUserAnonymous = false;

  @observable
  isMobile = false;

  initialQuery = qs.parse(window.location.search);

  @observable
  selectedAlertReceiveChannel?: AlertReceiveChannel['id'];

  @observable
  isLess1280: boolean;

  @observable
  features?: { [key: string]: boolean };

  @observable
  incidentFilters: any;

  @observable
  incidentsPage: any = this.initialQuery.p ? Number(this.initialQuery.p) : 1;

  // --------------------------

  userStore: UserStore = new UserStore(this);
  cloudStore: CloudStore = new CloudStore(this);
  grafanaTeamStore: GrafanaTeamStore = new GrafanaTeamStore(this);
  alertReceiveChannelStore: AlertReceiveChannelStore = new AlertReceiveChannelStore(this);
  outgoingWebhookStore: OutgoingWebhookStore = new OutgoingWebhookStore(this);
  alertReceiveChannelFiltersStore: AlertReceiveChannelFiltersStore = new AlertReceiveChannelFiltersStore(this);
  escalationChainStore: EscalationChainStore = new EscalationChainStore(this);
  escalationPolicyStore: EscalationPolicyStore = new EscalationPolicyStore(this);
  teamStore: TeamStore = new TeamStore(this);
  telegramChannelStore: TelegramChannelStore = new TelegramChannelStore(this);
  slackStore: SlackStore = new SlackStore(this);
  slackChannelStore: SlackChannelStore = new SlackChannelStore(this);
  heartbeatStore: HeartbeatStore = new HeartbeatStore(this);
  maintenanceStore: MaintenanceStore = new MaintenanceStore(this);
  scheduleStore: ScheduleStore = new ScheduleStore(this);
  userGroupStore: UserGroupStore = new UserGroupStore(this);
  alertGroupStore: AlertGroupStore = new AlertGroupStore(this);
  resolutionNotesStore: ResolutionNotesStore = new ResolutionNotesStore(this);
  apiTokenStore: ApiTokenStore = new ApiTokenStore(this);
  OrganizationLogStore: OrganizationLogStore = new OrganizationLogStore(this);
  globalSettingStore: GlobalSettingStore = new GlobalSettingStore(this);
  // stores

  async updateBasicData() {
    this.userStore.loadCurrentUser();
    this.teamStore.loadCurrentTeam();
    this.grafanaTeamStore.updateItems();
    this.updateFeatures();
    // this.userStore.updateItems();
    // this.maintenanceStore.updateMaintenances();
    this.userStore.updateNotificationPolicyOptions();
    this.userStore.updateNotifyByOptions();
    this.alertReceiveChannelStore.updateAlertReceiveChannelOptions();
    this.alertReceiveChannelStore.updateAlertReceiveChannelOptions();
    this.escalationPolicyStore.updateWebEscalationPolicyOptions();
    this.escalationPolicyStore.updateEscalationPolicyOptions();
    this.escalationPolicyStore.updateNumMinutesInWindowOptions();
  }

  async getUserRole() {
    const user = await getBackendSrv().get('/api/user');
    const userRoles = await getBackendSrv().get('/api/user/orgs');
    const userRole = userRoles.find(
      (userRole: { name: string; orgId: number; role: string }) => userRole.orgId === user.orgId
    );
    return userRole.role;
  }

  async finishSync(get_sync_response: any) {
    if (!get_sync_response.token_ok) {
      this.initializationError = 'OnCall was not able to connect back to this Grafana';
      return;
    }
    this.backendVersion = get_sync_response.version;
    this.backendLicense = get_sync_response.license;
    this.appLoading = false;
  }

  handleSyncException(e: any) {
    this.initializationError = e.response.status;
  }

  async startSync(key?: string) {
    try {
      return await startPluginSync();
    } catch (e) {
      if (e.response.status === 403) {
        this.correctProvisioningForInstallation = false;
        return;
      } else {
        this.initializationError = e.response.status;
        return;
      }
    }
  }

  resetStatusToDefault() {
    this.appLoading = true;
    this.pluginIsInitialized = true;
    this.correctProvisioningForInstallation = true;
    this.correctRoleForInstallation = true;
    this.signupAllowedForPlugin = true;
    this.initializationError = '';
    this.retrySync = false;
    this.isUserAnonymous = false;
  }

  async setupPlugin(meta: AppPluginMeta<OnCallAppSettings>) {
    this.resetStatusToDefault();

    if (!meta.jsonData?.onCallApiUrl) {
      this.pluginIsInitialized = false;
      return;
    }

    let syncStartStatus = await this.startSync();
    if (syncStartStatus.is_user_anonymous) {
      this.isUserAnonymous = true;
      return;
    } else if (!syncStartStatus.is_installed) {
      if (!syncStartStatus.allow_signup) {
        this.signupAllowedForPlugin = false;
        return;
      }
      const userRole = await this.getUserRole();
      if (userRole !== 'Admin') {
        this.correctRoleForInstallation = false;
        return;
      }
      await installPlugin();
    }

    let counter = 0;
    const interval = setInterval(() => {
      counter++;

      getPluginSyncStatus()
        .then((get_sync_response) => {
          if (get_sync_response.hasOwnProperty('token_ok')) {
            clearInterval(interval);
            this.finishSync(get_sync_response);
          }
        })
        .catch((e) => {
          clearInterval(interval);
          this.handleSyncException(e);
        });

      if (counter >= 5) {
        clearInterval(interval);
        this.retrySync = true;
      }
    }, 2000);
  }

  isUserActionAllowed(action: UserAction) {
    return this.userStore.currentUser && this.userStore.currentUser.permissions.includes(action);
  }

  hasFeature(feature: string | AppFeature) {
    // todo use AppFeature only
    return this.features?.[feature];
  }

  @observable
  async updateFeatures() {
    const response = await makeRequest('/features/', {});
    this.features = response.reduce(
      (acc: any, key: string) => ({
        ...acc,
        [key]: true,
      }),
      {}
    );
  }

  @action
  async removeSlackIntegration() {
    await this.slackStore.removeSlackIntegration();
  }

  @action
  async installSlackIntegration() {
    await this.slackStore.installSlackIntegration();
  }

  async getApiUrlForSettings() {
    const settings = await getBackendSrv().get('/api/plugins/grafana-oncall-app/settings');
    return settings.jsonData?.onCallApiUrl;
  }
}
