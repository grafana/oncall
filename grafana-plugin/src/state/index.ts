import { RootBaseStore } from './rootBaseStore';

export class RootStore extends RootBaseStore {}

export const rootStore = new RootStore();
