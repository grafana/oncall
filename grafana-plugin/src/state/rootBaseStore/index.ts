import { action, observable } from 'mobx';
import moment from 'moment-timezone';
import qs from 'query-string';
import { OnCallAppPluginMeta } from 'types';

import { AlertReceiveChannelStore } from 'models/alert_receive_channel/alert_receive_channel';
import { AlertReceiveChannel } from 'models/alert_receive_channel/alert_receive_channel.types';
import { AlertReceiveChannelFiltersStore } from 'models/alert_receive_channel_filters/alert_receive_channel_filters';
import { AlertGroupStore } from 'models/alertgroup/alertgroup';
import { ApiTokenStore } from 'models/api_token/api_token';
import { CloudStore } from 'models/cloud/cloud';
import { DirectPagingStore } from 'models/direct_paging/direct_paging';
import { EscalationChainStore } from 'models/escalation_chain/escalation_chain';
import { EscalationPolicyStore } from 'models/escalation_policy/escalation_policy';
import { GlobalSettingStore } from 'models/global_setting/global_setting';
import { GrafanaTeamStore } from 'models/grafana_team/grafana_team';
import { HeartbeatStore } from 'models/heartbeat/heartbeat';
import { MaintenanceStore } from 'models/maintenance/maintenance';
import { OrganizationLogStore } from 'models/organization_log/organization_log';
import { OutgoingWebhookStore } from 'models/outgoing_webhook/outgoing_webhook';
import { ResolutionNotesStore } from 'models/resolution_note/resolution_note';
import { ScheduleStore } from 'models/schedule/schedule';
import { SlackStore } from 'models/slack/slack';
import { SlackChannelStore } from 'models/slack_channel/slack_channel';
import { TeamStore } from 'models/team/team';
import { TelegramChannelStore } from 'models/telegram_channel/telegram_channel';
import { Timezone } from 'models/timezone/timezone.types';
import { UserStore } from 'models/user/user';
import { UserGroupStore } from 'models/user_group/user_group';
import { makeRequest } from 'network';
import { AppFeature } from 'state/features';
import PluginState from 'state/plugin';
import { isUserActionAllowed, UserActions } from 'utils/authorization';

// ------ Dashboard ------ //

export class RootBaseStore {
  @observable
  appLoading = true;

  @observable
  currentTimezone: Timezone = moment.tz.guess() as Timezone;

  @observable
  backendVersion = '';

  @observable
  backendLicense = '';

  @observable
  initializationError = null;

  @observable
  isMobile = false;

  initialQuery = qs.parse(window.location.search);

  @observable
  selectedAlertReceiveChannel?: AlertReceiveChannel['id'];

  @observable
  isLess1280: boolean;

  @observable
  features?: { [key: string]: boolean };

  @observable
  incidentFilters: any;

  @observable
  incidentsPage: any = this.initialQuery.p ? Number(this.initialQuery.p) : 1;

  @observable
  onCallApiUrl: string;

  // --------------------------

  userStore: UserStore = new UserStore(this);
  cloudStore: CloudStore = new CloudStore(this);
  directPagingStore: DirectPagingStore = new DirectPagingStore(this);
  grafanaTeamStore: GrafanaTeamStore = new GrafanaTeamStore(this);
  alertReceiveChannelStore: AlertReceiveChannelStore = new AlertReceiveChannelStore(this);
  outgoingWebhookStore: OutgoingWebhookStore = new OutgoingWebhookStore(this);
  alertReceiveChannelFiltersStore: AlertReceiveChannelFiltersStore = new AlertReceiveChannelFiltersStore(this);
  escalationChainStore: EscalationChainStore = new EscalationChainStore(this);
  escalationPolicyStore: EscalationPolicyStore = new EscalationPolicyStore(this);
  teamStore: TeamStore = new TeamStore(this);
  telegramChannelStore: TelegramChannelStore = new TelegramChannelStore(this);
  slackStore: SlackStore = new SlackStore(this);
  slackChannelStore: SlackChannelStore = new SlackChannelStore(this);
  heartbeatStore: HeartbeatStore = new HeartbeatStore(this);
  maintenanceStore: MaintenanceStore = new MaintenanceStore(this);
  scheduleStore: ScheduleStore = new ScheduleStore(this);
  userGroupStore: UserGroupStore = new UserGroupStore(this);
  alertGroupStore: AlertGroupStore = new AlertGroupStore(this);
  resolutionNotesStore: ResolutionNotesStore = new ResolutionNotesStore(this);
  apiTokenStore: ApiTokenStore = new ApiTokenStore(this);
  OrganizationLogStore: OrganizationLogStore = new OrganizationLogStore(this);
  globalSettingStore: GlobalSettingStore = new GlobalSettingStore(this);
  // stores

  async updateBasicData() {
    const updateFeatures = async () => {
      await this.updateFeatures();

      // Only fetch cloud connection status when cloud connection feature is enabled on OSS instance
      // Note that this.hasFeature can only be called after this.updateFeatures()
      if (this.hasFeature(AppFeature.CloudConnection)) {
        await this.cloudStore.loadCloudConnectionStatus();
      }
    };

    return Promise.all([
      this.teamStore.loadCurrentTeam(),
      this.grafanaTeamStore.updateItems(),
      updateFeatures(),
      this.userStore.updateNotificationPolicyOptions(),
      this.userStore.updateNotifyByOptions(),
      this.alertReceiveChannelStore.updateAlertReceiveChannelOptions(),
      this.escalationPolicyStore.updateWebEscalationPolicyOptions(),
      this.escalationPolicyStore.updateEscalationPolicyOptions(),
      this.escalationPolicyStore.updateNumMinutesInWindowOptions(),
    ]);
  }

  setupPluginError(errorMsg: string) {
    this.appLoading = false;
    this.initializationError = errorMsg;
  }

  /**
   * First check to see if the plugin has been provisioned (plugin's meta jsonData has an onCallApiUrl saved)
   * If not, tell the user they first need to configure/provision the plugin.
   *
   * Otherwise, get the plugin connection status from the OnCall API and check a few pre-conditions:
   * - plugin must be considered installed by the OnCall API
   * - token_ok must be true
   *   - This represents the status of the Grafana API token. It can be false in the event that either the token
   *   hasn't been created, or if the API token was revoked in Grafana.
   * - user must be not "anonymous" (this is determined by the plugin-proxy)
   * - the OnCall API must be currently allowing signup
   * - the user must have an Admin role
   * If these conditions are all met then trigger a data sync w/ the OnCall backend and poll its response
   * Finally, try to load the current user from the OnCall backend
   */
  async setupPlugin(meta: OnCallAppPluginMeta) {
    this.appLoading = true;
    this.initializationError = null;
    this.onCallApiUrl = meta.jsonData?.onCallApiUrl;

    if (!this.onCallApiUrl) {
      // plugin is not provisioned
      return this.setupPluginError('ðŸš« Plugin has not been initialized');
    }

    // at this point we know the plugin is provionsed
    const pluginConnectionStatus = await PluginState.checkIfPluginIsConnected(this.onCallApiUrl);
    if (typeof pluginConnectionStatus === 'string') {
      return this.setupPluginError(pluginConnectionStatus);
    }

    const { allow_signup, is_installed, is_user_anonymous, token_ok } = pluginConnectionStatus;
    if (is_user_anonymous) {
      return this.setupPluginError(
        'ðŸ˜ž Unfortunately Grafana OnCall is available for authorized users only, please sign in to proceed.'
      );
    } else if (!is_installed || !token_ok) {
      if (!allow_signup) {
        return this.setupPluginError('ðŸš« OnCall has temporarily disabled signup of new users. Please try again later.');
      }

      if (!isUserActionAllowed(UserActions.PluginsInstall)) {
        return this.setupPluginError(
          'ðŸš« An Admin in your organization must sign on and setup OnCall before it can be used'
        );
      }

      try {
        /**
         * this will install AND sync the necessary data
         * the sync is done automatically by the /plugin/install OnCall API endpoint
         * therefore there is no need to trigger an additional/separate sync, nor poll a status
         */
        await PluginState.installPlugin();
      } catch (e) {
        return this.setupPluginError(PluginState.getHumanReadableErrorFromOnCallError(e, this.onCallApiUrl, 'install'));
      }
    } else {
      const syncDataResponse = await PluginState.syncDataWithOnCall(this.onCallApiUrl);

      if (typeof syncDataResponse === 'string') {
        return this.setupPluginError(syncDataResponse);
      }

      // everything is all synced successfully at this point..
      this.backendVersion = syncDataResponse.version;
      this.backendLicense = syncDataResponse.license;
    }

    try {
      await this.userStore.loadCurrentUser();
    } catch (e) {
      return this.setupPluginError('OnCall was not able to load the current user. Try refreshing the page');
    }

    this.appLoading = false;
  }

  hasFeature(feature: string | AppFeature) {
    // todo use AppFeature only
    return this.features?.[feature];
  }

  @observable
  async updateFeatures() {
    const response = await makeRequest('/features/', {});
    this.features = response.reduce(
      (acc: any, key: string) => ({
        ...acc,
        [key]: true,
      }),
      {}
    );
  }

  @action
  async removeSlackIntegration() {
    await this.slackStore.removeSlackIntegration();
  }

  @action
  async installSlackIntegration() {
    await this.slackStore.installSlackIntegration();
  }

  async getApiUrlForSettings() {
    const settings = await PluginState.getGrafanaPluginSettings();
    return settings.jsonData?.onCallApiUrl;
  }
}
