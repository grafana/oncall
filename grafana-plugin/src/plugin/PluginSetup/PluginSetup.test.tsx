import React from 'react';

import * as runtime from '@grafana/runtime';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { RootBaseStore } from 'state/rootBaseStore/RootBaseStore';
import { useStore as useStoreOriginal } from 'state/useStore';

import { PluginSetup, PluginSetupProps } from './PluginSetup';

jest.mock('state/useStore');

jest.mock('@grafana/runtime', () => ({
  __esModule: true,

  config: {
    featureToggles: {
      topnav: undefined,
    },
  },

  getBackendSrv: () => ({
    get: jest.fn(),
    post: jest.fn(),
  }),

  PluginPage: (props: any) => <>{props.children}</>,
}));

jest.mock('grafana/app/core/core', () => ({
  contextSrv: {
    user: {
      orgRole: null,
    },
    hasAccess: (_action, _fallback): boolean => null,
  },
}));

const createComponentAndMakeAssertions = async (rootBaseStore: RootBaseStore) => {
  // mocks
  const mockedSetupPlugin = jest.fn();
  rootBaseStore.setupPlugin = mockedSetupPlugin;
  (useStoreOriginal as jest.Mock<ReturnType<typeof useStoreOriginal>>).mockReturnValue(rootBaseStore);

  // test setup
  const MockedInitializedComponent = jest.fn().mockReturnValue(<div>hello</div>);

  const props = {
    meta: {
      jsonData: 'hello',
    },
    InitializedComponent: MockedInitializedComponent,
  } as unknown as PluginSetupProps;

  const component = render(<PluginSetup {...props} />);

  // assertions
  expect(mockedSetupPlugin).toHaveBeenCalledTimes(1);
  expect(mockedSetupPlugin).toHaveBeenCalledWith(props.meta);
  expect(component.container).toMatchSnapshot();

  return mockedSetupPlugin;
};

describe('PluginSetup', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test('app is loading', async () => {
    const rootBaseStore = new RootBaseStore();
    await createComponentAndMakeAssertions(rootBaseStore);
  });

  test('there is an error message', async () => {
    const rootBaseStore = new RootBaseStore();
    rootBaseStore.initializationError = 'ohhhh noo';
    await createComponentAndMakeAssertions(rootBaseStore);
  });

  test('there is an error message - retry setup', async () => {
    const rootBaseStore = new RootBaseStore();
    rootBaseStore.initializationError = 'ohhhh noo';

    const mockedSetupPlugin = await createComponentAndMakeAssertions(rootBaseStore);

    await userEvent.click(screen.getByText('Retry'));

    expect(mockedSetupPlugin).toHaveBeenCalledTimes(2);
  });

  test('currently undergoing maintenance', async () => {
    const rootBaseStore = new RootBaseStore();
    rootBaseStore.currentlyUndergoingMaintenance = true;
    rootBaseStore.initializationError = 'there is some sort of maintenance';
    await createComponentAndMakeAssertions(rootBaseStore);
  });

  test('app successfully initialized', async () => {
    const rootBaseStore = new RootBaseStore();
    rootBaseStore.initializationError = null;
    await createComponentAndMakeAssertions(rootBaseStore);
  });

  test.each([true, false])('app initialized with topnavbar = %s', async (isTopNavBar: boolean) => {
    runtime.config.featureToggles.topnav = isTopNavBar;

    const rootBaseStore = new RootBaseStore();
    await createComponentAndMakeAssertions(rootBaseStore);
  });
});
