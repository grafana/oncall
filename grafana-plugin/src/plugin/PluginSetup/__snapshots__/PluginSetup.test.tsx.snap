// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluginSetup app initialized with topnavbar = false 1`] = `
<div>
  <div>
    hello
  </div>
</div>
`;

exports[`PluginSetup app initialized with topnavbar = true 1`] = `
<div>
  <div>
    hello
  </div>
</div>
`;

exports[`PluginSetup app is loading 1`] = `
<div>
  <div>
    hello
  </div>
</div>
`;

exports[`PluginSetup app successfully initialized 1`] = `
<div>
  <div>
    hello
  </div>
</div>
`;

exports[`PluginSetup currently undergoing maintenance 1`] = `
<div>
  <div
    class="spin"
  >
    <img
      alt="Grafana OnCall Logo"
      src="[object Object]"
    />
    <div
      class="spin-text"
    >
      there is some sort of maintenance
    </div>
  </div>
</div>
`;

exports[`PluginSetup there is an error message - retry setup 1`] = `
<div>
  <div
    class="spin"
  >
    <img
      alt="Grafana OnCall Logo"
      src="[object Object]"
    />
    <div
      class="spin-text"
    >
      ohhhh noo
    </div>
    <div
      class="configure-plugin"
    >
      <div
        class="css-ffyaiw-horizontal-group"
        style="width: 100%; height: 100%;"
      >
        <div
          class="css-18qv8yz-layoutChildrenWrapper"
        >
          <button
            class="css-72lnkn-button"
            type="button"
          >
            <span
              class="css-1riaxdn"
            >
              Retry
            </span>
          </button>
        </div>
        <div
          class="css-18qv8yz-layoutChildrenWrapper"
        >
          <a
            class="css-72lnkn-button"
            href="/plugins/grafana-oncall-app?page=configuration"
            tabindex="0"
          >
            <span
              class="css-1riaxdn"
            >
              Configure Plugin
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PluginSetup there is an error message 1`] = `
<div>
  <div
    class="spin"
  >
    <img
      alt="Grafana OnCall Logo"
      src="[object Object]"
    />
    <div
      class="spin-text"
    >
      ohhhh noo
    </div>
    <div
      class="configure-plugin"
    >
      <div
        class="css-ffyaiw-horizontal-group"
        style="width: 100%; height: 100%;"
      >
        <div
          class="css-18qv8yz-layoutChildrenWrapper"
        >
          <button
            class="css-72lnkn-button"
            type="button"
          >
            <span
              class="css-1riaxdn"
            >
              Retry
            </span>
          </button>
        </div>
        <div
          class="css-18qv8yz-layoutChildrenWrapper"
        >
          <a
            class="css-72lnkn-button"
            href="/plugins/grafana-oncall-app?page=configuration"
            tabindex="0"
          >
            <span
              class="css-1riaxdn"
            >
              Configure Plugin
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
